tools/SetupFlow/DevHome.SetupFlow.ElevatedComponent/ElevatedComponentOperation.cs

```diff
@@ -34,37 +52,190 @@ public void WriteToStdOut(string value)
+        where TTaskArguments : ITaskArguments
+    {
+        try
+        {
+            ValidateAndBeginOperation(taskArguments);
+            var result = await executeFunction();
+            var success = resultProcessorFunction(result);
+            EndOperation(taskArguments, success);
+            return result;
+        }
+        catch (Exception e)
+        {
+            Log.Logger?.ReportError(Log.Component.Elevated, $"Failed to validate or execute operation", e);
+            EndOperation(taskArguments, false);
+            throw;
+        }
+    }
+
+    /// <summary>
+    /// Validate the execution of the operation with the provided tasks arguments
+    /// </summary>
+    /// <param name="taskArguments">Task arguments for the operation to execute</param>
+    /// <exception cref="InvalidOperationException">Thrown if this operation cannot be performed</exception>
+    private void ValidateAndBeginOperation(ITaskArguments taskArguments)
+    {
+        lock (_operationStateLock)
+        {
+            // Check if this operation has already been attempted before.
+            if (_operationsState.TryGetValue(taskArguments, out var operationState))
+            {
+                if (operationState.InProgress)
+                {
+                    throw new InvalidOperationException($"Failed to perform operation because an identical operation is still executing.");
+                }
+
+                if (operationState.RemainingAttempts <= 0)
+                {
+                    throw new InvalidOperationException($"Failed to perform operation because the maximum number of attempts has been reached.");
+                }
+
+                operationState.RemainingAttempts--;
+                operationState.InProgress = true;
+            }
+            else
+            {
+                // This is the first time this operation is being executed
+                _operationsState.Add(taskArguments, new OperationState
+                {
+                    RemainingAttempts = MaxRetryAttempts,
+                    InProgress = true,
+                });
+            }
+        }
+    }
+
+    /// <summary>
+    /// End operation with a boolean indicating whether it was successful or not.
+    /// </summary>
+    /// <param name="taskArguments">Task arguments </param>
+    /// <param name="isSuccessful">Boolean indicating whether the operation was successful or not</param>
+    private void EndOperation(ITaskArguments taskArguments, bool isSuccessful)
+    {
+        lock (_operationStateLock)
+        {
+            if (_operationsState.TryGetValue(taskArguments, out var operationState))
+            {
+                if (isSuccessful)
+                {
+                    // Since the operation succeeded, prevent further attempts to
+                    // execute it again in the lifetime of the elevated process.
+                    operationState.RemainingAttempts = 0;
+                }
```

Comment:
Decrementing the attempt is done when first starting the operation in `ValidateAndBeginOperation()`. But to answer your question, yes, and when successful we explicitly set the remaining attempts to zero to prevent unexpected behavior.